### PR TITLE
Descise complecity in schema

### PR DESCRIPTION
**Pull request by berenda-andrey**

### CODE REVIEW

File: winter/schema/docstring_parser.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import docstring_parser
```
Comment (berenda-andrey) [inline winter/schema/docstring_parser.py]: For depending not much on this library

File: winter/schema/docstring_parser.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    def get_description(self, argument_name: str, invalid_type_hint: bool = False):
+        param_doc = self.params_docs.get(argument_name)
+        description = param_doc.description if param_doc is not None else ''
+
+        if invalid_type_hint:
+            description = self.update_doc_with_invalid_hype_hint(description)
+        return description
+
+    def update_doc_with_invalid_hype_hint(self, doc: str):
```
Comment (mofr) [inline winter/schema/docstring_parser.py]: This method is not related to docstring at all, please move it out

File: winter/schema/docstring_parser.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    def __init__(self, doc: str):
+        docstring = docstring_parser.parse(doc)
+        self.params_docs = {param_doc.arg_name: param_doc for param_doc in docstring.params}
+
+    def get_description(self, argument_name: str, invalid_type_hint: bool = False):
+        param_doc = self.params_docs.get(argument_name)
+        description = param_doc.description if param_doc is not None else ''
+
+        if invalid_type_hint:
+            description = self.update_doc_with_invalid_hype_hint(description)
```
Comment (mofr) [inline winter/schema/docstring_parser.py]: Please do not call it here, because it's unexpected behavior for the caller.

File: winter/schema/docstring_parser.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import docstring_parser
+
+
+class DocstringParser:
+
+    def __init__(self, doc: str):
+        docstring = docstring_parser.parse(doc)
+        self.params_docs = {param_doc.arg_name: param_doc for param_doc in docstring.params}
+
+    def get_description(self, argument_name: str, invalid_type_hint: bool = False):
```
Comment (mofr) [inline winter/schema/docstring_parser.py]: get_description -> get_argument_description
Because docstring contains not only arguments description.

File: winter/core/component_method_argument.py [hunk 5]
```diff
@@ -16,3 +19,14 @@ class ComponentMethodArgument:
...
         return self.method.signature.parameters[self.name]
+
+    def has_default(self) -> bool:
+        return self.default is not inspect.Parameter.empty
+
+    @cached_property
+    def default(self):
+        return self.parameter.default
+
+    @property
```
[the remark this answers is not in the document]
Comment (mofr) [reply winter/core/component_method_argument.py]: Yes, given class is frozen, we can add a cache to all heavy properties.

File: winter/schema/docstring_parser.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import docstring_parser
+
+
+class DocstringParser:
```
Comment (mofr) [inline winter/schema/docstring_parser.py]: Consider renaming it to `Docstring` because only constructor does parsing, and the class basically represents already parsed docstring structure.

File: winter/schema/query_parameters_inspector.py [hunk 7]
```diff
@@ -1,23 +1,18 @@
...
 from ..query_parameter import get_query_param_mapping
 
 
 class QueryParametersInspector(MethodArgumentsInspector):
 
     def inspect_parameters(self, method: ComponentMethod) -> List[openapi.Parameter]:
-        docstring = docstring_parser.parse(method.func.__doc__)
-        params_docs = {param_doc.arg_name: param_doc for param_doc in docstring.params}
-
+        docstring_parser = DocstringParser(method.func.__doc__)
```
Comment (mofr) [inline winter/schema/query_parameters_inspector.py]: Please consider the following alternative: `docstring = method.docstring`. It will allow to parse docstring for each method only once.

File: tests/test_query_parameter.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import typing
```
Comment (berenda-andrey) [inline tests/test_query_parameter.py]: I need to add it because this code is not covered but I need to change it

File: winter/core/utils.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# noinspection PyPep8Naming
+class cached_property:
```
Comment (berenda-andrey) [inline winter/core/utils.py]: I added copy from Django
For less depending on it

File: tests/test_query_parameter.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+from django.http import HttpRequest
+from django.http import QueryDict
+from rest_framework.exceptions import ParseError
+from rest_framework.request import Request
+
+import winter
+from winter.query_parameter import QueryParameterResolver
+
+
+def get_resolver():
```
Comment (mofr) [inline tests/test_query_parameter.py]: What this function for? Maybe just inline it?

File: winter/core/docstring.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import docstring_parser
+
+
+class Docstring:
+
+    def __init__(self, doc: str):
+        docstring = docstring_parser.parse(doc)
+        self.params_docs = {param_doc.arg_name: param_doc for param_doc in docstring.params}
+
+    def get_argument_description(self, argument_name: str) -> bool:
```
Comment (mofr) [inline winter/core/docstring.py]: Wrong return type hint `bool`.

**Comment by berenda-andrey** [comment]
Not finished yet

File: winter/schema/query_parameters_inspector.py [hunk 12]
```diff
@@ -1,54 +1,56 @@
...
+            default = argument.default if argument.has_default() else None
 
-            param_doc = params_docs.get(query_parameter_name)
-            description = param_doc.description if param_doc else ''
-            default_params = {}
-            default = argument.parameter.default
-            if default is not inspect.Parameter.empty:
-                default_params['default'] = default
-            required = default is inspect.Parameter.empty and not type_utils.is_optional(argument.type_)
             type_info = get_argument_type_info(argument)
```
Comment (mofr) [inline winter/schema/query_parameters_inspector.py]: Maybe, include description and default_type_info to `get_argument_type_info`?

Comment (berenda-andrey) [reply winter/schema/query_parameters_inspector.py]: I set default_type_info to string
If we need to can change it in future

File: winter/schema/query_parameters_inspector.py [hunk 13]
```diff
@@ -1,54 +1,48 @@
...
+if typing.TYPE_CHECKING:  # pragma: no cover
+    from ..routing import Route
 
-class QueryParametersInspector(MethodArgumentsInspector):
 
-    def inspect_parameters(self, method: ComponentMethod) -> List[openapi.Parameter]:
-        docstring = docstring_parser.parse(method.func.__doc__)
-        params_docs = {param_doc.arg_name: param_doc for param_doc in docstring.params}
+class QueryParametersInspector(MethodArgumentsInspector):
+    _default_type_info = TypeInfo(openapi.TYPE_STRING)
```
Comment (mofr) [inline winter/schema/query_parameters_inspector.py]: It's not used anymore

File: winter/schema/query_parameters_inspector.py [hunk 14]
```diff
@@ -1,54 +1,48 @@
...
-            required = default is inspect.Parameter.empty and not type_utils.is_optional(argument.type_)
-            type_info = get_argument_type_info(argument)
+            default = argument.default if argument.has_default() else None
 
-            if type_info is None:
-                type_info_data = {'type': openapi.TYPE_STRING}
-                description = update_doc_with_invalid_hype_hint(description)
-            else:
-                type_info_data = type_info.as_dict()
+            type_info_data = get_argument_type_info(argument)
```
Comment (mofr) [inline winter/schema/query_parameters_inspector.py]: Maybe rename `get_argument_type_info` to something `get_argument_info` to avoid messing up with TypeInfo. And assign the result to something similar (not type info, because it's not TypeInfo, more likely parameter_data)

File: winter/schema/query_parameters_inspector.py [hunk 15]
```diff
@@ -1,54 +1,48 @@
...
 
-            param_doc = params_docs.get(query_parameter_name)
-            description = param_doc.description if param_doc else ''
-            default_params = {}
-            default = argument.parameter.default
-            if default is not inspect.Parameter.empty:
-                default_params['default'] = default
-            required = default is inspect.Parameter.empty and not type_utils.is_optional(argument.type_)
-            type_info = get_argument_type_info(argument)
+            default = argument.default if argument.has_default() else None
```
Comment (mofr) [inline winter/schema/query_parameters_inspector.py]: Consider including it to `get_argument_type_info` as well.
